### PR TITLE
refactor: defaulting REST options

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #13361 

Add an internal function for defaulting REST options (and use it).

We are still calling this thing in the stub factory which feels wrong to me, but this is a step in the right direction.

Add unit tests that verify the defaulting logic (assuming we call `PopulateRestOptions()`, that is).

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13363)

### CODE REVIEW

File: google/cloud/internal/populate_rest_options.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include "google/cloud/options.h"
+#include "google/cloud/version.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+/**
+ * Modify @p opts to have default values for REST options.
```
Comment (devbww) [inline google/cloud/internal/populate_rest_options.h:27]: Nit: "Modify" seems like an inappropriate verb for a function that takes and returns by value.

File: google/cloud/internal/populate_rest_options.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * It is expected that `EndpointOption` has already been set on `opts` by
+ * `PopulateCommonOptions`.
+ *
+ * @param opts the current options. Any values already present in this
+ *     collection are not modified.
```
Comment (devbww) [inline google/cloud/internal/populate_rest_options.h:33]: This doesn't appear to be true when a present `EndpointOption` doesn't start with "http".

File: google/cloud/internal/populate_rest_options.cc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+  if (!opts.has<rest_internal::LongrunningEndpointOption>()) {
+    opts.set<rest_internal::LongrunningEndpointOption>(
+        "https://longrunning.googleapis.com");
+  }
+  if (opts.has<EndpointOption>()) {
+    std::string endpoint = opts.get<EndpointOption>();
+    if (!absl::StartsWithIgnoreCase(endpoint, "http")) {
+      opts.set<EndpointOption>(absl::StrCat("https://", endpoint));
+    }
```
Comment (devbww) [inline google/cloud/internal/populate_rest_options.cc:40]: Nit: I know this is only capturing existing behavior, but starting with "http" is not the same thing as starting with "http://" or "https://".

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/populate_rest_options.h (hunk 1 above)
Comment (dbolduc) [reply google/cloud/internal/populate_rest_options.h]: Done.

File: google/cloud/internal/populate_rest_options.h (hunk 2 above)
Comment (dbolduc) [reply google/cloud/internal/populate_rest_options.h]: Fixed, thanks.

File: google/cloud/internal/populate_rest_options.cc (hunk 3 above)
Comment (dbolduc) [reply google/cloud/internal/populate_rest_options.cc:40]: Ack, I am going to defer addressing this comment in this PR.

Comment (dbolduc) [reply google/cloud/internal/populate_rest_options.cc:40]: #13397

**Pull request merged**